[JENKINS-69113] renovate progressBar

mawinter69 opened the pull request:
The progressbar used at various places is replaced by a div/span combination and gets a new styling that is in sync with other ui elements. It is a little bit higher, has rounded corners and some padding, the minwidth is set to 10% so we always have at least a bit to show, it also avoids that the border is not rounded, e.g. when we're at just 2% and the progressbar is only 2px wide.
Animation is now based on css and no longer a gif and can also be enabled via a flag.
The build progress bar shown on run and console views is now updated dynamically including the tooltip.
The progress bar used in progressive rendering is doubled in size to make it more prominent that it is still running (See [JENKINS-72138], this doesn't solve the problem but might lower the chance that people reload the mentioned page because they think nothing happens).
This also fixes JENKINS-68631

**After** 
Build history widget:
<img width="242" alt="image" src="https://github.com/jenkinsci/jenkins/assets/17767050/0dcd8d1c-ca0f-4725-b596-4bba0bf59238">

Executor widget:
<img width="255" alt="image" src="https://github.com/jenkinsci/jenkins/assets/17767050/c7c363af-9760-4a18-b751-3353569add81">

Run view:
Here the progress bar will change over time
<img width="613" alt="image" src="https://github.com/jenkinsci/jenkins/assets/17767050/c78dc407-5ecd-4e7c-a560-1a87f5af5926">
<img width="309" alt="image" src="https://github.com/jenkinsci/jenkins/assets/17767050/7ba2547e-1c63-4425-b77d-196cecb3b459">


Console view with stuck:
<img width="608" alt="image" src="https://github.com/jenkinsci/jenkins/assets/17767050/ee84e5c4-b658-4fc3-b973-e0677cfda08c">

Progressive rendering (agent build history) with bigger progress bar:
<img width="597" alt="image" src="https://github.com/jenkinsci/jenkins/assets/17767050/2c067807-45e7-419f-bfaf-d3c7436b7016">




<!-- Comment:
A great PR typically begins with the line below.
Replace XXXXX with the numeric part of the issue ID you created in Jira.
Note that if you want your changes backported into LTS, you need to create a Jira issue. See https://www.jenkins.io/download/lts/#backporting-process for more information.
-->

See [JENKINS-69113](https://issues.jenkins.io/browse/JENKINS-69113).

<!-- Comment:
If the issue is not fully described in Jira, add more information here (justification, pull request links, etc.).

 * We do not require Jira issues for minor improvements.
 * Bug fixes should have a Jira issue to facilitate the backporting process.
 * Major new features should have a Jira issue.
-->

### Testing done
Manual testing:
- create job that runs for 1 minute
- trigger job (no previous builds) -> progress bar is at 100% and animated (on executor widgets, run page and console page)
- trigger job (with previous builds) -> progress bar is animated and increases over time

To see stuck jobs
- Create job that does nothing
- run the job several times
- Change job to sleep for 1 minute
- run job -> the progressbar should now be red animated.

Tested for freestyle and pipeline jobs

<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->

### Proposed changelog entries

- JENKINS-69113, renovate progressBar

<!-- Comment:
The changelog entry should be in the imperative mood; e.g., write "do this"/"return that" rather than "does this"/"returns that".
For examples, see: https://www.jenkins.io/changelog/

Remove JENKINS-XXXXX if there is no issue for the pull request.

You may add multiple changelog entries if applicable by adding a new entry to the list, e.g.
- JENKINS-123456, First changelog entry
- Second changelog entry
-->

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/core-pr-reviewers.
-->

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

mawinter69 committed:
[JENKINS-69113] renovate progressBar

The progressbar used at various places is replaced by a div/span
compbination and gets a new styling that is in sync with other ui
elements. Mainly it has rounded corners.
The bar is always animated opposed to old behaviour where it was only
animated when the estimated remaining time was unknown. Animation is now
based on css and no longer a gif.
All colors are defined via variables so they can be adjusted by themes.
The build progress bar shown on run and console views is now updated
dynamically.
The progress bar used in progressive rendering is doubled in size to
make it more prominent that it is still running (See [JENKINS-72138],
this doesn't solve the problem but might lower the chance that people
reload the mentioned page because they think nothing happens).

mawinter69 committed:
apply prettier

mawinter69 committed:
scss style

mawinter69 committed:
set status url

the parameters page also includes the buildCaption.jelly. But the js is
using a relative url thus accessing `statusIcon` fails.
Store status url in the div and read from there.

mawinter69 committed:
apply behaviour so tooltip is shown after icon update

fix url

janfaracik commented:
Looking good! 

I was looking at doing something similar last year, but never finished it. Here’s the branch if it’s at all helpful - https://github.com/jenkinsci/jenkins/compare/master...janfaracik:jenkins:progress-bar-new?expand=1.


https://github.com/jenkinsci/jenkins/assets/43062514/1981bd2d-370d-4ad7-ab6e-af49a016316c



A few things -

* I think it’d be best to use the existing colour palette for consistency rather than using the colours from the existing progress bar, e.g. using the accent colour for progress, error colour for errors etc.
* I think there should be separate animations for progress/indeterminate states, e.g. a soft pulse for progress, horizontal translation when indeterminate
* It’s a little chunky at the moment and draws a lot of attention - maybe that’s okay though?

Let me know your thoughts 👍

janfaracik commented on core/src/main/resources/lib/hudson/progressBar.jelly:
```diff
@@ -32,19 +32,20 @@ THE SOFTWARE.
 <?jelly escape-by-default='true'?>
 <j:jelly xmlns:j="jelly:core" xmlns:x="jelly:xml" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form">
-  <table class="progress-bar ${attrs.red?'red':null}" id="${attrs.id}" tooltip="${attrs.tooltip}" href="${attrs.href}" style="${attrs.href!=null ? 'cursor:pointer' : null}">
+  <div class="app-progress-bar ${attrs.large?'app-progress-bar-large':null} ${attrs.red?'red':null}"
```
Would be nice if the tag for the progress bar changed if it's a link - see https://github.com/janfaracik/jenkins/blob/8e88712bf2f7de338773ed03723c6682e5ef022a/core/src/main/resources/lib/hudson/progressBar.jelly#L44

mawinter69 replied:
done

NotMyFault commented:
> Looking good!
> 
> I was looking at doing something similar last year, but never finished it. Here’s the branch if it’s at all helpful - [`master...janfaracik:jenkins:progress-bar-new`?expand=1 (compare)](https://github.com/jenkinsci/jenkins/compare/master...janfaracik:jenkins:progress-bar-new?expand=1).
> 
>  Screen.Recording.2024-01-03.at.13.51.48.mov 
> A few things -
> 
> * I think it’d be best to use the existing colour palette for consistency rather than using the colours from the existing progress bar, e.g. using the accent colour for progress, error colour for errors etc.
> * I think there should be separate animations for progress/indeterminate states, e.g. a soft pulse for progress, horizontal translation when indeterminate
> * It’s a little chunky at the moment and draws a lot of attention - maybe that’s okay though?
> 
> Let me know your thoughts 👍

I'm a big fan of your proposed implementation of the progress bars. I agree we should stick with the current color palette for consistency :)

mawinter69 committed:
incorporate feedback

use existing colors
only animate when unknown or with flag
animate via build caption and progressive rendering

mawinter69 commented:
The animation is now only used in the widgets when the progress is unknown or is is set via an option.
Using existing color palette now

mawinter69 committed:
adjust class name

mawinter69 commented:
Should the padding be a bit bigger maybe?
like this: 
<img width="177" alt="image" src="https://github.com/jenkinsci/jenkins/assets/17767050/442dda7c-94e2-477b-ade4-b5e3cc17238f">

mawinter69 committed:
adjust bg color

mawinter69 committed:
fix style

mawinter69 committed:
sRGB

yaroslavafenkin commented on core/src/main/resources/lib/hudson/progressBar.jelly:
```diff
@@ -22,29 +22,48 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+  <x:element name="${attrs.href != null ? 'a' : 'div'}">
+    <x:attribute name="class">app-progress-bar ${attrs.large?'app-progress-bar--large':null} ${attrs.red?'app-progress-bar--error':null}</x:attribute>
+    <x:attribute name="href">${attrs.href}</x:attribute>
+    <x:attribute name="tooltip"><j:out value="${attrs.tooltip}"/></x:attribute>
```
What's the reason to have `j:out` here and 2 lines below? Do we expect HTML in these tooltips?

mawinter69 replied:
there is html in those tooltips, currently just a `<br>`, see https://github.com/jenkinsci/jenkins/blob/2923a7785ea735b0b90a8cfb2cbf67bbf864356e/core/src/main/resources/lib/hudson/buildProgressBar.properties#L23 and all the other languages.
I could replace all the `<br>`in the properties with `\n`, then it should work without the `j:out`

mawinter69 replied:
removed the `j:out`

mawinter69 committed:
avoid j:out

NotMyFault committed:
Merge branch 'master' into JENKINS-71795

NotMyFault reviewed (approved):
I'm really fond of the redesign, thanks!

timja reviewed (approved):
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

basil commented:
[Breaks `jenkins.branch.NoTriggerBranchPropertyTest#suppressNothing` and `jenkins.branch.NoTriggerBranchPropertyTest#suppressEvents`](https://github.com/jenkinsci/bom/pull/2936/checks?check_run_id=21793204796):

```
java.lang.AssertionError: 

Expected: is <2>
     but: was <3>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
	at jenkins.branch.NoTriggerBranchPropertyTest.suppressNothing(NoTriggerBranchPropertyTest.java:135)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:656)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

mawinter69 commented:
I wonder how this change could cause this error. This change only affect jelly/js files so it is pure UI

basil commented:
Are you implying that this PR may not be the cause of the test failures? The tests deterministically fail with this PR and pass when this PR is reverted, so this PR is the (proximate) cause of the test failures.

jglick commented:
I also got here via `git bisect`. `NoTriggerBranchPropertyTest` runs `JenkinsRule.configRoundtrip` so it does exercise GUI code.

CloudBees also encountered a couple of regressions in ATH `BuildHistoryTest` at https://github.com/jenkinsci/acceptance-test-harness/blob/0ac2b663128e2c20bfdcacbf095496bd7c6b43ab/src/main/java/org/jenkinsci/test/acceptance/po/BuildHistory.java#L31 which look like they could have been caused by this PR but I have not yet confirmed that.

mawinter69 commented:
It's definitely happening after configRoundTrip
But I think it is not related to this change.
When I add a 30 sec sleep after the configRoundTrip, then also the current master of branch-api plugin built against 2.426.1 fails.
Isn't there an automatic first run triggered after saving the configuration?

jglick commented:
Thanks, I can take a look at whether `NoTriggerBranchPropertyTest` is just overly sensitive to timing.
